nemo/collections/llm/peft/lora.py

```diff
@@ -54,6 +58,118 @@ def forward(self, x):
         return linear_output + adapter_output, bias
 
 
+class LoRAModuleConnector:
+    """Gets and sets attributes of a module that are relevant to LoRA."""
+
+    @staticmethod
+    def get_in_features(m: nn.Module):
+        """Gets the number of input features for the module.
+
+        Args:
+            m: The module to get the input features for.
+
+        Returns:
+            int: The number of input features.
+        """
+        if HAVE_TE:
+            from megatron.core.transformer.custom_layers.transformer_engine import (
+                TEColumnParallelLinear,
+                TERowParallelLinear,
+            )
+
+        if isinstance(m, ColumnParallelLinear) or isinstance(m, RowParallelLinear):
+            return m.input_size
+        elif HAVE_TE and isinstance(m, TEColumnParallelLinear):
+            return m.in_features
+        elif HAVE_TE and isinstance(m, TERowParallelLinear):
+            tp_size = parallel_state.get_tensor_model_parallel_world_size()
+            return m.in_features * tp_size
+        else:
+            raise ValueError(f"Unsupported module type for LoRA: {type(m)}")
+
+    @staticmethod
+    def get_out_features(m: nn.Module):
```

Comment:
Can we narrow the types here rather than use nn.Module?

 'ColumnParallelLinear' | 'RowParallelLinear' | 'TERowParallelLinear' | TERowParallelLinear'

Reply:
Updated